### PR TITLE
Merge improvements to environment scaling

### PR DESCRIPTION
**Pull request by kshyatt**
Going to squash and merge this after running some checks

### CODE REVIEW

**Comment by kshyatt** [comment]
There's a million commits on this which I intend to squash. Just checked that the energies come out right.

File: src/environments.jl [hunk 1]
```diff
@@ -4,79 +4,238 @@ struct Environments
...
+        @inbounds for row in order
+            # construct environment for the row, have to do this every time
+            @timeit "build row Env" begin
+                Env  = copy(Env_below)
+                for row_ in (row, row+1)
+                    tmp = A[row_, col]
+                    tmp *= ops[row_]
+                    tmp *= prev_mps[row_]
+                    Env *= tmp
+                    Env *= dag(prime(A[row_, col]))
```
Comment (emstoudenmire) [inline src/environments.jl:164]: I believe this is a D^6. However, if the guess MPS can be used here, the algorithm can be reduced to a D^5 until one builds the part with the two guess-MPS tensors missing. Is that this part?

File: src/environments.jl [hunk 2]
```diff
@@ -4,79 +4,238 @@ struct Environments
...
+    Envs_above[end]  = is_cu ? cuITensor(1.0) : ITensor(1.0)
+    @inbounds for (ii, env_row) in enumerate(reverse(order))
+        tmp = A[env_row, col]
+        tmp *= ops[env_row]
+        tmp *= prev_mps[env_row]
+        next_row = isodd(sweep) ? env_row + 1 : env_row - 1
+        Env_above  = next_row > 0 ? copy(Envs_above[next_row]) : dummy 
+        Env_above *= tmp
+        Env_above *= dag(prime(A[env_row, col]))
+        Env_above *= guess[env_row]
```
Comment (emstoudenmire) [inline src/environments.jl:78]: I believe switching the above two lines makes this function D^5 instead of D^6

File: src/environments.jl [hunk 3]
```diff
@@ -4,79 +4,238 @@ struct Environments
...
+                    tmp = A[row_, col]
+                    tmp *= ops[row_]
+                    tmp *= prev_mps[row_]
+                    Env *= tmp
+                    Env *= dag(prime(A[row_, col]))
+                    if row_ == row
+                        Env_below = copy(Env)
+                    end
+                end
+                Env *= Envs_above[row+2]
```
Comment (emstoudenmire) [inline src/environments.jl:169]: I believe this is a D^6 step, but one we can't avoid so it's ok.

**Review by emstoudenmire — CHANGES_REQUESTED** [review]
Looks good, except for the two comments I made about scaling. I think the one inside constructEnvsAbove is the most important one to check, regarding scaling. But please also take a look at the first comment in fitPEPSMPO to see if that D^6 can be avoided too. The very last D^6 before the SVD can't be avoided, I don't think.

File: src/environments.jl (hunk 1 above)
Comment (kshyatt) [reply src/environments.jl:164]: This is the part with the two tensors missing, yes

File: src/environments.jl (hunk 2 above)
Comment (kshyatt) [reply src/environments.jl:78]: Tried this and it actually made scaling worse! Reverted back to the old order.

Comment (emstoudenmire) [reply src/environments.jl:78]: Ok, good that you checked. We'll have to come back to some of this after doing timings for empirical scaling. But right now everything in this part of the code looks to be no more than D^6 and that's the most important thing!

**Review by emstoudenmire — APPROVED** [review]

**Pull request merged**